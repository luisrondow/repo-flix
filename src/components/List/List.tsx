import Text from '../Text'

import ArrowIcon from '../Icons/ArrowIcon'
import useHorizontalScroll from '../../hooks/useHorizontalScroll'
import { ListProps } from './List.types'

const LIST_HEIGHT = {
  listContainer: {
    medium: 'h-56',
    large: 'h-72',
  },
  arrowContainer: {
    medium: 'h-48',
    large: 'h-64',
  },
}

const List = (props: ListProps) => {
  const { size, title, children } = props

  const { listRef, isListLeftScrolled, handleScroll } = useHorizontalScroll()

  return (
    <div data-testid="repositories-list-container" className="relative mt-8">
      <Text as="h1" className="pl-16">
        {title}
      </Text>
      <div
        data-testid="repositories-list"
        ref={listRef}
        className={`${LIST_HEIGHT.listContainer[size]} no-scrollbar relative flex h-56 w-full snap-mandatory flex-row flex-nowrap items-center space-x-4 overflow-x-auto overflow-y-hidden py-4 pl-16 pr-4`}
      >
        {children}
      </div>
      <div
        className={`${LIST_HEIGHT.arrowContainer[size]} absolute bottom-0 right-0 z-20 mb-4 flex w-60 items-center justify-end bg-gradient-to-r from-transparent to-base opacity-0 transition-opacity hover:opacity-100`}
      >
        <button onClick={() => handleScroll('right')} className="h-full pl-8 pr-2">
          <ArrowIcon direction="right" />
        </button>
      </div>
      {isListLeftScrolled ? (
        <div
          className={`${LIST_HEIGHT.arrowContainer[size]} absolute bottom-0 left-0 z-20 mb-4 flex h-48 w-60 items-center justify-start bg-gradient-to-l from-transparent to-base opacity-0 transition-opacity hover:opacity-100`}
        >
          <button onClick={() => handleScroll('left')} className="h-full pr-8 pl-2">
            <ArrowIcon direction="left" />
          </button>
        </div>
      ) : null}
    </div>
  )
}

export default List
