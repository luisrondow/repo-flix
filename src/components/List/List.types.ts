export type ListProps = {
  title: string
  size: 'medium' | 'large'
  children: React.ReactNode
}
