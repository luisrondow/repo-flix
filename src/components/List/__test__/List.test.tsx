import { act, render } from '@testing-library/react'
import List from '../List'

// const repositories = [
//   {
//     id: '10270250',
//     name: 'facebook/react',
//     url: 'https://github.com/facebook/react',
//     image: 'https://opengraph.githubassets.com/123abc/facebook/react',
//   },
//   {
//     id: '126577260',
//     name: 'trekhleb/javascript-algorithms',
//     url: 'https://github.com/trekhleb/javascript-algorithms',
//     image: 'https://opengraph.githubassets.com/123abc/trekhleb/javascript-algorithms',
//   },
//   {
//     id: '2126244',
//     name: 'twbs/bootstrap',
//     url: 'https://github.com/twbs/bootstrap',
//     image: 'https://opengraph.githubassets.com/123abc/twbs/bootstrap',
//   },
//   {
//     id: '6498492',
//     name: 'airbnb/javascript',
//     url: 'https://github.com/airbnb/javascript',
//     image: 'https://opengraph.githubassets.com/123abc/airbnb/javascript',
//   },
//   {
//     id: '29028775',
//     name: 'facebook/react-native',
//     url: 'https://github.com/facebook/react-native',
//     image: 'https://opengraph.githubassets.com/123abc/facebook/react-native',
//   },
// ]

// jest.mock('../../../../../hooks/useRepositories', () => {
//   return {
//     __esModule: true,
//     default: () => ({
//       repositories,
//       isLoading: false,
//       isError: false,
//       error: null,
//     }),
//   }
// })

describe('List', () => {
  it('should render the correctly', () => {
    const { container, getByTestId, getByText, getAllByRole } = render(
      <List size="medium" title="Mock title">
        <div>Mock children</div>
        <div>Mock children</div>
        <div>Mock children</div>
      </List>,
    )

    expect(getByTestId('repositories-list-container')).toBeInTheDocument()
    expect(getByTestId('repositories-list')).toBeInTheDocument()
    expect(getByText('Mock title')).toBeInTheDocument()
    expect(getAllByRole('button')).toHaveLength(1)

    expect(container).toMatchSnapshot()
  })

  it('should scroll the list correctly', () => {
    const { getByTestId, getAllByRole } = render(
      <List size="medium" title="Mock title">
        <div>Mock children</div>
        <div>Mock children</div>
        <div>Mock children</div>
      </List>,
    )

    window.HTMLElement.prototype.scrollBy = jest.fn()

    const list = getByTestId('repositories-list')
    let buttons = getAllByRole('button')

    act(() => buttons[0].click())

    expect(list.scrollBy).toHaveBeenCalledWith({
      left: 448,
      behavior: 'smooth',
    })
    expect(getAllByRole('button')).toHaveLength(2)

    buttons = getAllByRole('button')

    act(() => buttons[1].click())

    expect(list.scrollBy).toHaveBeenCalledWith({
      left: -448,
      behavior: 'smooth',
    })
    expect(getAllByRole('button')).toHaveLength(1)
  })
})
