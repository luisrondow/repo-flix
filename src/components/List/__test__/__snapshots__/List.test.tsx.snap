// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render the correctly 1`] = `
<div>
  <div
    class="relative mt-8"
    data-testid="repositories-list-container"
  >
    <h1
      class="pl-16 text-[#e0def4] font-roboto text-4xl font-bold"
    >
      Mock title
    </h1>
    <div
      class="h-56 no-scrollbar relative flex h-56 w-full snap-mandatory flex-row flex-nowrap items-center space-x-4 overflow-x-auto overflow-y-hidden py-4 pl-16 pr-4"
      data-testid="repositories-list"
    >
      <div>
        Mock children
      </div>
      <div>
        Mock children
      </div>
      <div>
        Mock children
      </div>
    </div>
    <div
      class="h-48 absolute bottom-0 right-0 z-20 mb-4 flex w-60 items-center justify-end bg-gradient-to-r from-transparent to-base opacity-0 transition-opacity hover:opacity-100"
    >
      <button
        class="h-full pl-8 pr-2"
      >
        <svg
          class="h-12 w-12 stroke-gray-light"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.25 4.5l7.5 7.5-7.5 7.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
